### ISSUE TITLE
Request-based Fetch-and-Op and Compare-and-Swap

### ISSUE THREAD

**Issue by omor1** [mpi-forum/mpi-issues#128]
# Problem

Currently request-based RMA (§11.3.5) is defined for Put, Get, Accumulate, and Get-Accumulate. However, it is not defined for Fetch-and-Op and Compare-and-Swap.

# Proposal

Add `MPI_RFETCH_AND_OP` and `MPI_RCOMPARE_AND_SWAP`.

# Changes to the Text

#####  MPI_RFETCH_AND_OP(origin_addr, result_addr, datatype, target_rank, target_disp, op, win, request)

Type | Name | Description
----- | ------ | ------------
IN | origin_addr | initial address of buffer (choice)
OUT | result_addr | initial address of result buffer (choice)
IN | datatype | datatype of the entry in origin, result, and target buffers (handle)
IN | target_rank | rank of target (non-negative integer)
IN | target_disp | displacement from start of window to beginning of target buffer (non-negative integer)
IN | op | reduce operation (handle)
IN | win | window object (handle)
OUT | request | RMA request (handle)

```
int MPI_Rfetch_and_op(const void *origin_addr, void *result_addr,
             MPI_Datatype datatype, int target_rank, MPI_Aint target_disp,
             MPI_Op op, MPI_Win win, MPI_Request *request)

MPI_Rfetch_and_op(origin_addr, result_addr, datatype, target_rank,
             target_disp, op, win, request, ierror)
    TYPE(*), DIMENSION(..), INTENT(IN), ASYNCHRONOUS ::  origin_addr
    TYPE(*), DIMENSION(..), ASYNCHRONOUS ::  result_addr
    TYPE(MPI_Datatype), INTENT(IN) ::  datatype
    INTEGER, INTENT(IN) ::  target_rank
    INTEGER(KIND=MPI_ADDRESS_KIND), INTENT(IN) ::  target_disp
    TYPE(MPI_Op), INTENT(IN). ::  op
    TYPE(MPI_Win), INTENT(IN) ::  win
    TYPE(MPI_Request), INTENT(OUT) ::  request
    INTEGER, OPTIONAL, INTENT(OUT) ::  ierror

MPI_RFETCH_AND_OP(ORIGIN_ADDR, RESULT_ADDR, DATATYPE, TARGET_RANK,
             TARGET_DISP, OP, WIN, REQUEST, IERROR)
    <type> ORIGIN_ADDR(*), RESULT_ADDR(*)
    INTEGER(KIND=MPI_ADDRESS_KIND) TARGET_DISP
    INTEGER DATATYPE, TARGET_RANK, OP, WIN, REQUEST, IERROR
```

`MPI_RFETCH_AND_OP` is similar to `MPI_FETCH_AND_OP` (Section 11.3.4), except that it allocates a communication request object and associates it with the request handle (the argument `request`) that can be used to wait or test for completion. The completion of an `MPI_RFETCH_AND_OP` operation indicates that the data is available in the result buffer and the origin buffer is free to be updated. It does not indicate that the operation has been completed at the target window.

##### MPI_RCOMPARE_AND_SWAP(origin_addr, compare_addr, result_addr, datatype, target_rank, target_disp, win, request)

Type | Name | Description
----- | ------ | ------------
IN | origin_addr | initial address of buffer (choice)
IN | compare_addr | initial address of compare buffer (choice)
OUT | result_addr | initial address of result buffer (choice)
IN | datatype | datatype of the element in all buffers (handle)
IN | target_rank | rank of target (non-negative integer)
IN | target_disp | displacement from start of window to beginning of target buffer (non-negative integer)
IN | win | window object (handle)
OUT | request | RMA request (handle)

```
int MPI_Rcompare_and_swap(const void *origin_addr, const void *compare_addr,
             void *result_addr, MPI_Datatype datatype, int target_rank,
             MPI_Aint target_disp, MPI_Win win)

MPI_Rcompare_and_swap(origin_addr, compare_addr, result_addr, datatype,
             target_rank, target_disp, win, ierror)
    TYPE(*), DIMENSION(..), INTENT(IN), ASYNCHRONOUS ::  origin_addr
    TYPE(*), DIMENSION(..), INTENT(IN), ASYNCHRONOUS ::  compare_addr
    TYPE(*), DIMENSION(..), ASYNCHRONOUS ::  result_addr
    TYPE(MPI_Datatype), INTENT(IN) ::  datatype
    INTEGER, INTENT(IN) ::  target_rank
    INTEGER(KIND=MPI_ADDRESS_KIND), INTENT(IN) ::  target_disp
    TYPE(MPI_Win), INTENT(IN) ::  win
    TYPE(MPI_Request), INTENT(OUT) ::  request
    INTEGER, OPTIONAL, INTENT(OUT) ::  ierror

MPI_RCOMPARE_AND_SWAP(ORIGIN_ADDR, COMPARE_ADDR, RESULT_ADDR, DATATYPE,
             TARGET_RANK, TARGET_DISP, WIN, IERROR)
    <type> ORIGIN_ADDR(*), COMPARE_ADDR(*), RESULT_ADDR(*)
    INTEGER(KIND=MPI_ADDRESS_KIND) TARGET_DISP
    INTEGER DATATYPE, TARGET_RANK, WIN, REQUEST, IERROR
```

`MPI_RCOMPARE_AND_SWAP` is similar to `MPI_COMPARE_AND_SWAP` (Section 11.3.4), except that it allocates a communication request object and associates it with the request handle (the argument `request`) that can be used to wait or test for completion. The completion of an `MPI_RCOMPARE_AND_SWAP` operation indicates that the data is available in the result buffer and the origin and compare buffers are free to be updated. The completion also indicates that the operation has been completed at the target window; comparison of the compare and result buffers is sufficient to determine if the operation succeeded.

# Impact on Implementations

Implementations will need to implement request-based Fetch-and-Op and Compare-and-Swap. For the former, this can be easily done via the request-based Get-Accumulate, but implementation of the latter is more difficult.

For Compare-and-Swap, the completion of the request must also indicate that the operation has completed at the target window, unlike the other request-based RMA operations. This is because CAS is frequently used to implement mutual exclusion; a process must be able to determine if the compare-and-swap was successful or if it must be retried.

# Impact on Users

None

# References

None

**Comment by jeffhammond** [comment]
I’m strongly in favor of NB CAS. Not convinced we need NB FOP but it looks
stupid if FOP is the only RMA comm that doesn’t have NB variant.

Thanks for this proposal.
-- 
Jeff Hammond
jeff.science@gmail.com
http://jeffhammond.github.io/

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@jeffhammond I've never thought of request-based RMA functions as being nonblocking operations - or, at least, no more nonblocking than the RMA function without requests.

MPI_PUT is already nonblocking (because writing to the origin buffer before a subsequent synchronisation is not permitted; it is a conflicting access) but there is no way to discover completion of the individual operation, only in batch at the next synchronisation point.
MPI_RPUT does not, therefore, denote a nonblocking version of MPI_PUT; it provides an individualised completion detection mechanism for a put operation.

Are the CAS and FOP functions blocking? That is, do they enforce the guarantee that the resources specified by the arguments can be reused when the call returns? In particular, can the user immediately scribble over the memory at origin_addr and read the correct value from result_addr?

Looking at the definition in MPI-3.1, there is a small ambiguity on that point. The operation will be atomic when it happens, yes; but do I need to synchronise the window before reading memory at result_addr? I guess the intent is that these functions are blocking; i.e. that they involve a round-trip to the target and will return only once the result value from the target has been written locally at result_addr.

@omor1 If CAS & FOP and currently blocking, and the new functions will be nonblocking, then the convention is to use "I" for immediate/incomplete, rather than "R" as the prefix.

Of course, RMA is already not consistent with this naming convention - a user could reasonably expect, from the names alone, that MPI_PUT is blocking and MPI_RPUT should have been called MPI_IPUT. This would be a bad change to make for many reasons. Those reasons might also be a good justification for naming nonblocking CAS & FOP operations differently, i.e. using "I" as the prefix - it warns the user that these two name/functionality shifts are not equivalent.

**Comment by omor1** [comment]
All MPI RMA calls are nonblocking, with bulk completion (§11.7.3):
> RMA calls must have local semantics, except when required for synchronization with other RMA calls.

 For active target synchronization, `MPI_WIN_FENCE` and `MPI_WIN_COMPLETE` will complete calls at the origin; `MPI_WIN_FENCE` and `MPI_WIN_WAIT` will complete calls at the target. For passive target synchronization, calls complete at the origin with `MPI_WIN_UNLOCK`, `MPI_WIN_UNLOCK_ALL`, `MPI_WIN_FLUSH`, `MPI_WIN_FLUSH_ALL`, `MPI_WIN_FLUSH_LOCAL`, and `MPI_WIN_FLUSH_LOCAL_ALL`. The first four calls also additionally complete the calls at the target.

As you said, what the request-based calls allow is local completion of individual calls instead of in bulk. As far as I can tell, if remote completion is additionally required, an explicit call to `MPI_WIN_FLUSH` would be needed afterwards.

It is my interpretation that CAS and FOP are nonblocking, like all RMA operations; one cannot use the result values until the call completes, at least not while being portable and correct. This is supported by example 11.20, which has an explicit call to `MPI_WIN_FLUSH` after `MPI_COMPARE_AND_SWAP` to ensure completion.

I also want to note that the current specification for `MPI_RGET_ACCUMULATE` (which I copied to `MPI_RFETCH_AND_OP`) can result in rather curious behavior. Consider the following:
```
const int value = 1;
int result[2];
MPI_Request req[2];
MPI_Win window = ...

MPI_Win_lock_all(0, window);
/* get current value  and accumulate 1 into offset 0 of the window at rank 0 */
MPI_Rget_accumulate(&value, 1, MPI_INT, &result[0], 1, MPI_INT,
                    0, 0, 1, MPI_INT, MPI_SUM, window, &req[0]);
/* do it again */
MPI_Rget_accumulate(&value, 1, MPI_INT, &result[1], 1, MPI_INT,
                    0, 0, 1, MPI_INT, MPI_SUM, window, &req[1]);
/* complete calls */
MPI_Waitall(2, req, MPI_STATUSES_IGNORE);
assert(result[0] != result[1] && "How could this happen?");
MPI_Win_unlock_all(window);
```

Even assuming no other process has any overlapping writes, stories, or accumulates to the same location in the window, the assertion can still fail. That's because the specification states the following (emphasis mine):
> The completion of an `MPI_RGET_ACCUMULATE` operation indicates that the data is available in the result buffer and the origin buffer is free to be updated. **It does not indicate that the operation has been completed at the target window.**

This seems to me to be nonintuitive and incorrect. I doubt any implementation actually has this behavior, but that it may nonetheless occur seems wrong to me—doubly so for FOP and CAS. For CAS in particular, this is the reason I specified that completion of the request also completes it at the target. It may be that it's reasonable to have two associated requests—one that completes the call at the origin and another that completes it at the target.

What's the current behavior of `MPI_WIN_FLUSH_LOCAL` and `MPI_WIN_FLUSH_LOCAL_ALL` with CAS? The specification states:
> For example, after this routine completes, the user may reuse any buffers provided to put, get, or accumulate operations.
Does this mean that the input arguments can be reused but output ones cannot? Or is the entire operation complete at this point, with similar semantics to get (for which it does seem to be the case)?

**Comment by hjelmn** [comment]
I don't see how the assertion can fail in this case. The first accumulate will return one less than the second one due to the accumulate ordering on the window. waw is part of the default (https://www.mpi-forum.org/docs/mpi-3.1/mpi31-report/node291.htm).

**Comment by omor1** [comment]
The operations are ordered, yes—the first get-accumulate will commit at the target before the second. However, the completion of the get-accumulate is only at the origin, not at the target—it is not required to have actually done the accumulation before the current value is returned. This would appear to conflict with the requirement that the operation is atomic. The point is that the specification is ambiguous about this.

A better example might use `MPI_REPLACE`:
```
const int value[2] = {1, 2};
int result[2];
MPI_Request req[2];
/* fetch-and-set */
MPI_Rget_accumulate(&value[0], 1, MPI_INT, &result[0], 1, MPI_INT,
                    0, 0, 1, MPI_INT, MPI_REPLACE, window, &req[0]);
/* do it again */
MPI_Rget_accumulate(&value[1], 1, MPI_INT, &result[1], 1, MPI_INT,
                    0, 0, 1, MPI_INT, MPI_SUM, window, &req[1]);
/* complete calls */
MPI_Waitall(2, req, MPI_STATUSES_IGNORE)
```

You can have `result[0] == result[1]`, but once the operations is completed at the target (via `MPI_Win_unlock_all`), a subsequent get would show the value to be 2 rather than 1. Again, this depends on interpretation on the atomicity of the operation as compared to local versus remote completion.

**Comment by hjelmn** [comment]
True but my interpretation of atomicity (the one used in Open MPI) would require that the first operation complete at the target before the second operation could possibly be complete locally. That is the only sensible interpretation as otherwise the operations are not "atomic".

**Comment by hjelmn** [comment]
Otherwise what is the point of MPI_OP_NO_OP? If your interpretation is correct there is no difference between MPI_Get_accumulate with MPI_OP_NO_OP and MPI_Get.

**Comment by hjelmn** [comment]
Now, when you weaken the accumulate ordering then things definitely get ambiguous.

**Comment by omor1** [comment]
I agree, which is why I think it makes no sense to state that completion of `MPI_RGET_ACCUMULATE` does not mean that the operation completed at the target. In any case, that's a bit off-track for this issue—I'll open up another to discuss the semantics of local completion versus remote completion for accumulate functions.

**Comment by hjelmn** [comment]
@omor1 yes. ok. that is true. i think the wording is a bit weird for sure.

**Comment by hjelmn** [comment]
Also, I don't think this PR is a bad idea. At the very least from a completeness perspective.

**Comment by pavanbalaji** [comment]
I agree with this proposal for the CAS part.  For Fetch-and-op, this was discussed during MPI-3 (@hjelmn was also a part of the discussion, but perhaps he forgot the discussion).  The reason it was not added was because one can always use `MPI_Rget_accumulate` if you want a request-based operation.  The only reason FOP exists, together with `MPI_Get_accumulate`, is because the MPI implementation does not have to check a few parameters (e.g., count is always `1` and the datatype is always basic), which can improve performance for very fast networks.

**Comment by omor1** [comment]
> The only reason FOP exists, together with `MPI_Get_accumulate`, is because the MPI implementation does not have to check a few parameters (e.g., count is always 1 and the datatype is always basic), which can improve performance for very fast networks.

Out of curiosity, why then does no equivalent for `MPI_ACCUMULATE` exist? I've encountered a couple instances where one wants to accumulate into a single value, but doesn't care to retrieve it.

```
int MPI_Do_op(const void *origin_addr, MPI_Datatype datatype,
             int target_rank, MPI_Aint target_disp,
             MPI_Op op, MPI_Win win)
```

For consistency's sake I'd like to see a request-based fetch-and-op, with a note stating that it may be no more efficient than the equivalent `MPI_RGET_ACCUMULATE`. It adds to the API surface, but implementations can trivially implement it.

**Comment by pavanbalaji** [comment]
> Out of curiosity, why then does no equivalent for MPI_ACCUMULATE exist? I've encountered a couple instances where one wants to accumulate into a single value, but doesn't care to retrieve it.

We added functions based on use cases.  There weren't any use cases pointed out for an equivalent of `MPI_ACCUMULATE`.  But if you have one, we can certainly consider it for MPI-4.

> For consistency's sake I'd like to see a request-based fetch-and-op, with a note stating that it may be no more efficient than the equivalent MPI_RGET_ACCUMULATE. It adds to the API surface, but implementations can trivially implement it.

I'm all for consistency, but there are several instances in the standard where we have not been consistent.  And several of the Forum members (primarily MPI implementors) have been against adding functions just for the sake of consistency without having strong use cases for it.  This is mainly because adding, testing, and maintaining such functionality takes time and effort.  You can propose it and see what the Forum members think of it.

Btw, one bad example where the consistency argument worked terribly is when `MPI_Cancel` was added to send operations.  There was no use case for it, but because the MPI standard allows it, we had to implement it.  It was extremely hard to implement and test and get it right, and in some cases it even hurt performance for the non-cancel path.  Your function is not that extreme, but I just wanted to give an example where things went really badly when the only argument for adding something was consistency.

**Comment by omor1** [comment]
Fair enough, makes sense.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@omor1 I just want to point out that your reasoning for "all RMA operations are nonblocking" is a sequitur.
> All MPI RMA calls are nonblocking, with bulk completion (§11.7.3):
>> RMA calls must have local semantics, except when required for synchronization with other RMA calls.

Local/non-local and blocking/non-blocking are orthogonal concepts.
Example of local and blocking: MPI_BSEND
Example of local and nonblocking: MPI_IBSEND

Thus, "RMA calls must have local semantics" does not imply "RMA calls must be nonblocking".

Local/non-local relates to whether return of the function depends on other MPI processes calling MPI functions.
Blocking/nonblocking relates to whether return of the function permits the user to re-use the parameters supplied to the function.

I do not yet see anything in the MPI Standard that prevents some RMA functions being blocking and some being nonblocking. In fact, the text about CAS and FOP strongly suggests to me that they are local blocking functions. They do not depend on remote MPI function calls (this property called is local in MPI terms; this does not imply nonblocking in the MPI sense of that term, even though it fits with the English phrase "does not block"). The user can(re-)use all the parameters when the function returns (this property is called blocking in MPI terms).

This is a further example of why the Terms & Conventions WG is suggesting clarifications of the definitions of these terms. Specifically, we are suggesting that we stop referring to MPI functions as blocking or nonblocking and reserve that terminology for MPI operations. Instead, we should use complete/incomplete to refer to MPI functions that do/do not permit the user to re-use all parameters at the point of return.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@omor1 Example 11.20 does not seem to be good support for the argument that CAS is supposed to be nonblocking.

That example is confusing because its pseudo-code syntax does not match the actual function definitions (the return value of the function is not the result, the function takes the address of the result as a parameter, allowing the possibility of it being an OUT parameter updated during the function or an address that MPI takes ownership of until some sort of later completion). The caption for that example only talks about what happens before the MPI_BARRIER, not what happens afterwards, so we cannot be sure why the MPI_WIN_SYNC in the loop is present.

The MPI_WIN_SYNC function ensures that the public and private copies of a specified window are synchronised. It does not ensure completion, either local or remote, of any RMA operations (that is done by MPI_WIN_FLUSH and its variants).

The result buffer does not have to be in window memory. If it is not window memory, then MPI_WIN_SYNC has no effect on it at all. Even if it were, all accesses in the example are load/stores to the private copy - so the value in the public copy is not relevant to the correctness of the example. The MPI_WIN_SYNC inside the loop only ensures that the public and private copies of A are synchronised, which is not important because all accesses to A (except the initialisation ```A=0```) are to the public copy via MPI functions.

This suggests to me that the MPI_WIN_SYNC inside the loop is superfluous and could be removed without affecting the correctness of the example. Without the superfluous MPI_WIN_SYNC call the argument that this example supports the case for CAS being nonblocking falls apart.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@omor1 [removed previous comment that mistakenly argued about a MPI_WIN_SYNC inside the loop, which is actually MPI_WIN_FLUSH - apologies]

**Comment by hjelmn** [comment]
For reference MPI_Fetch_and_op and MPI_Compare_and_swap are both blocking in the osc/rdma implementation in Open MPI if the hardware supports the atomic.

**Comment by omor1** [comment]
@dholmes-epcc-ed-ac-uk right, I misread the specification. The only thing that's stated is that the operation isn't guaranteed to be complete at the origin until the next synchronization call, whether that's `MPI_WIN_FENCE`, `MPI_WIN_COMPLETE`, `MPI_WIN_FLUSH`, `MPI_WIN_FLUSH_ALL`, `MPI_WIN_FLUSH_LOCAL`, `MPI_WIN_FLUSH_LOCAL_ALL`, `MPI_WIN_UNLOCK`, `MPI_WIN_UNLOCK_ALL`, or completion of the request (for request-based RMA). This gives freedom to the implementation to have the call complete immediately or delay it until later (blocking or non-blocking).

**Comment by pavanbalaji** [comment]
I’m pretty sure we say in the MPI standard that all RMA communication calls are nonblocking (that is, they must return in a finite amount of time irrespective of the status of other processes in the system).  I’ll look up the actual wording once I get to my laptop, but I remember we argued about it and decided to leave the MPI-2 wording in that already said that.

**Comment by omor1** [comment]
> The user can(re-)use all the parameters when the function returns (this property is called blocking in MPI terms).

Where do you see this? The only thing written about CAS is
> This function compares one element of type datatype in the compare buffer `compare_addr` with the buffer at offset `target_disp` in the target window specified by `target_rank` and `win` and replaces the value at the target with the value in the origin buffer `origin_addr` if the compare buffer and the target buffer are identical. The original value at the target is returned in the buffer `result_addr`. The parameter `datatype` must belong to one of the following categories of predefined datatypes: C integer, Fortran integer, Logical, Multi-language types, or Byte as specified in Section 5.9.2. The origin and result buffers (`origin_addr` and `result_addr`) must be disjoint.

This is similar language as used for other RMA calls—nothing about it indicates to me that the operation is required to be blocking (i.e. parameters can be re-used).

**Comment by pavanbalaji** [comment]
> For reference MPI_Fetch_and_op and MPI_Compare_and_swap are both blocking in the osc/rdma implementation in Open MPI if the hardware supports the atomic.

FWIW, this is not actually “blocking” in MPI semantics because it still completes in a finite amount of time (assuming you already finished the epoch opening synchronization, such as lock or fence).  It’s still nonblocking.

**Comment by pavanbalaji** [comment]
With respect to my previous comment about all RMA communication calls being nonblocking, this is explicitly stated in the MPI-3.1 standard in section 11.3 (page 417, lines 12-20).

**Comment by omor1** [comment]
@pavanbalaji good catch!

As a side note, what are the progress semantics for the request-based RMA when using `MPI_TEST` in a loop? I assume the intent is that it's similar to nonblocking two-sided communication, i.e. that the call eventually becomes enabled and then completes, but it's not clear to me whether this is strictly required by the specification. Unlike in §3.7.4, which explicitly states that calling `MPI_TEST` in a loop will eventually complete the request if a matching send or receive has been started, §11.7.3 does not state this—only that the communication must complete once it is enabled, and that it is enabled at latest at the ensuing synchronization call. Obviously `MPI_WAIT` must enable a request-based RMA communication, but it's not clear if `MPI_TEST` does the same.

**Comment by pavanbalaji** [comment]
`MPI_TEST` has the same semantics for request-based RMA operations too.

**Comment by jeffhammond** [comment]
To be clear, my first comment (https://github.com/mpi-forum/mpi-issues/issues/128#issuecomment-480151315) was dumb and has been edited appropriately.  RMA is nonblocking and has local completion semantics in a proper implementation.

**Comment by pavanbalaji** [comment]
> To be clear, my first comment ([#128 (comment)](https://github.com/mpi-forum/mpi-issues/issues/128#issuecomment-480151315)) was dumb and has been edited appropriately. RMA is nonblocking and has local completion semantics in a proper implementation.

RMA communication operations themselves do not have immediate local completion.  You need to do `MPI_Win_flush_local` or `MPI_Test/MPI_Wait` (for request-based operations) to get that.

**Comment by omor1** [comment]
Currently the pull request has both FOP and CAS, but FOP can be easily removed.

**Comment by jeffhammond** [comment]
I recall Jeremiah Willcock asked for request-based FOP and CAS in 2014, for consistency purposes.  I think forcing users to use RGETACC because we refuse to add RFOP is unreasonable.  Having consistency in the standard is important.  And what if MPI_Request allocation isn't always expensive compared to FOP?

**Comment by Wee-Free-Scot** [comment]
> And what if MPI_Request allocation isn't always expensive compared to FOP?

Request allocation isn't expensive when its a persistent request and it is reused (started & completed) enough times to amortize the cost. Oh wait, we don't have persistent RMA yet, do we? ;)

**Comment by jeffhammond** [comment]
One thing that complicates this topic is that local completion of a fetching atomic does NOT imply remote completion.  I learned this the hard way in ARMCI-MPI.  It is counterintuitive (to me, at least) but the only way to remote-complete a fetching atomic is with a flush.

The non-equivalence of request-based and flushed completion of fetching atomics may need to be addressed along with this issue.
